kongswap_adaptor/src/deposit/test_failed_transfer_from.rs

```diff
@@ -0,0 +1,708 @@
+        symbol: symbol_1,
+        ledger_fee_decimals: Nat::from(FEE_ICP),
+    };
+
+    let owner_account = sns_treasury_manager::Account {
+        owner: Principal::from_text("2vxsx-fae").unwrap(),
+        subaccount: None,
+    };
+
+    thread_local! {
+        static MEMORY_MANAGER: RefCell<MemoryManager<DefaultMemoryImpl>> =
+            RefCell::new(MemoryManager::init(DefaultMemoryImpl::default()));
+
+        static BALANCES: RefCell<StableBalances> =
+            MEMORY_MANAGER.with(|memory_manager|
+                RefCell::new(
+                    StableCell::init(
+                        memory_manager.borrow().get(BALANCES_MEMORY_ID),
+                        ConfigState::default()
+                    )
+                    .expect("BALANCES init should not cause errors")
+                )
+            );
+
+        static AUDIT_TRAIL: RefCell<StableAuditTrail> =
+            MEMORY_MANAGER.with(|memory_manager|
+                RefCell::new(
+                    StableVec::init(
+                        memory_manager.borrow().get(AUDIT_TRAIL_MEMORY_ID)
+                    )
+                    .expect("AUDIT_TRAIL init should not cause errors")
+                )
+            );
+    }
+
+    let amount_0_decimals = 500 * E8;
+    let amount_1_decimals = 400 * E8;
+    let allowances = vec![
+        // SNS
+        Allowance {
+            asset: asset_0.clone(),
+            owner_account,
+            amount_decimals: Nat::from(amount_0_decimals),
+        },
+        // ICP
+        Allowance {
+            asset: asset_1.clone(),
+            owner_account,
+            amount_decimals: Nat::from(amount_1_decimals),
+        },
+    ];
+
+    let mock_agent = MockAgent::new(*SELF_CANISTER_ID)
+        .add_call(
+            sns_ledger,
+            make_approve_request(amount_0_decimals, FEE_SNS),
+            Ok(Nat::from(amount_0_decimals)),
+        )
+        .add_call(
+            icp_ledger,
+            make_approve_request(amount_1_decimals, FEE_ICP),
+            Ok(Nat::from(amount_1_decimals)),
+        )
+        .add_call(
+            sns_ledger,
+            make_balance_request(*SELF_CANISTER_ID),
+            Nat::from(amount_0_decimals - FEE_SNS),
+        )
+        .add_call(
+            icp_ledger,
+            make_balance_request(*SELF_CANISTER_ID),
+            Nat::from(amount_1_decimals - FEE_ICP),
+        )
+        .add_call(
+            *KONG_BACKEND_CANISTER_ID,
+            make_add_token_request(token_0.clone()),
+            Ok(make_add_token_reply(
+                1,
+                "IC".to_string(),
+                sns_id,
+                "My DAO Token".to_string(),
+                "DAO".to_string(),
+                FEE_SNS,
+            )),
+        )
+        .add_call(
+            *KONG_BACKEND_CANISTER_ID,
+            make_add_token_request(token_1.clone()),
+            Ok(make_add_token_reply(
+                2,
+                "IC".to_string(),
```

Comment:
This could be hidden in the helper, as it's confusing (we're only dealing with IC tokens in this canister anyway)